ocaml/xapi/xapi_pif_helpers.ml

```diff
@@ -255,10 +255,29 @@ let is_device_underneath_same_type ~__context pif1 pif2 =
   in
   get_device_info pif1 = get_device_info pif2
 
+let get_non_link_ipv6 ~__context ~pif =
+  let non_link_ip6s =
+    List.map
+      (fun ip6 ->
+        match Ipaddr.V6.Prefix.of_string ip6 with
+        | Error _ ->
+            None
+        | Ok ip6 ->
+            let ip6 = Ipaddr.V6.Prefix.address ip6 in
+            if Ipaddr.V6.scope ip6 = Ipaddr.Link then
+              None
+            else
+              Some (Ipaddr.V6.to_string ip6)
+      )
+      (Db.PIF.get_IPv6 ~__context ~self:pif)
+  in
+  List.map (Option.fold ~none:"" ~some:Fun.id) non_link_ip6s
+  |> List.filter (( <> ) "")
```

Comment:
There's no need to make so many transformations to filter out invalid values, a single List.filter_map is enough.
The applied function can also be expressed using a series of transformations:
```suggestion
  let valid_nonlink ipv6 =
    let open Ipaddr.V6 in
    ipv6
    |> Prefix.of_string
    |> Result.to_option
    |> Fun.flip Option.bind @@ fun cidr ->
       let addr = Prefix.address cidr in
       match scope addr with Ipaddr.Link -> None | _ -> Some (to_string addr)
  in
  List.filter_map valid_nonlink (Db.PIF.get_IPv6 ~__context ~self:pif)
```

Reply:
Thx for the suggestion, its way simpler indeed. i'll try it and report here. :)